### PR TITLE
[core] Fix memleaks in the callback

### LINKED ISSUE #46952: Release test long_running_many_drivers.aws failed
Release test **long_running_many_drivers.aws** failed. See https://buildkite.com/ray-project/release/builds/20440#01911608-011f-4fb0-b5fc-a815fcef4a8c for more details.

Managed by OSS Test Policy

### LINKED ISSUE #46954: Release test long_running_many_jobs.aws failed
Release test **long_running_many_jobs.aws** failed. See https://buildkite.com/ray-project/release/builds/20440#01911608-01ea-46e0-bdfa-6b91e936fa90 for more details.

Managed by OSS Test Policy

### PR DESCRIPTION
**Pull request by rynewang**
Refcounting in Cython is really pesky, because Cython makes a bunch of + and - waltz and we need to play along. Per Cython document, the C++ functor PyCallback calls `converter_` and gets a PyObject, should be responsible to DECREF after it's out of scope, but we didn't. Fixes by adding the DECREF.

Fixes #46954.
Fixes #46952.

### CODE REVIEW

File: python/ray/includes/gcs_client.pxi [hunk 1]
```diff
@@ -589,6 +575,23 @@ cdef void assign_and_decrement_fut(result, void* fut_ptr) with gil:
...
+        #    a borrowed reference, and the callee does not do anything with it.
+        # 3. (PyCallback) when a func defines a local `object = f()`, the object keeps
+        #    the new reference, until it goes out of scope when the function should
+        #    DECREF it.
+        #
+        # The problem is our (3) is happening in the C++ PyCallback class, where we
+        # don't want to do any DECREFing, since we want most things happen in Cython.
+        # Hence we have to do the DECREFing here. After this DECREF(result), nobody
+        # should touch the result.
+        cpython.Py_DECREF(result)
```
Comment (jjyao) [inline python/ray/includes/gcs_client.pxi:592]: let's move this to c++

**Review by jjyao — APPROVED** [review]

File: python/ray/includes/gcs_client.pxi [hunk 2]
```diff
@@ -589,6 +575,23 @@ cdef void assign_and_decrement_fut(result, void* fut_ptr) with gil:
         else:
             fut.set_result(ret)
     finally:
+        # Refcounting in Cython is really pesky, because Cython makes a bunch of + and -
```
Comment (jjyao) [inline python/ray/includes/gcs_client.pxi:578]: Also add a link to the cython doc where the info is from

**Comment by jjyao** [comment]
Also update the PR description to match the code

File: python/ray/includes/gcs_client.pxi (hunk 2 above)
Comment (rynewang) [reply python/ray/includes/gcs_client.pxi]: added in cpp

**Pull request merged**